# Data Management & Analytics - Project 1

### 1. Work instructions

**Due date**: April 26 2017 (00:00 pm) <br>
**Deliverable**: Python notebook to be sent at "loic.winckelmans@ulb.ac.be". No additional documents <br>
**Quotations**: 31 questions - 1 point per question - 15% of final quote <br>
**Group composition**: Groups of 2 or 3 students <br>
**Q&A**: Q&A session organised on April 14 2017

**Group composition**:

Student 1:  
Student 2:  
Student 3:  

### 2. Context

You recently heard that more than 33.000 people die from gun-death in the U.S. every year. <br>

As a data scientist you want to use data acquisition techniques and databases to acquire a deeper understanding of the subject. <br> 

You came across a .CSV dataset from the CDC (https://www.cdc.gov/) that includes information about gun-death in the US in the years 2012-2014. <br>

You then decided to download the file (guns.csv) and explore the data. <br>

### 3. Data

The dataset includes data regarding the victim's age, sex, race, education, intent, time (month and year) and place of death, and whether or not police was at the place of death.

- year: The year in which the fatality occurred, integer <br>

- month: The month in which the fatality occurred, integer <br>

- intent: The intent of the perpetrator of the crime. This can be Suicide, Accidental, NA, Homicide, or Undetermined, string

- police: Whether a police officer was involved with the shooting. Either 0 (false) or 1 (true), integer

- sex: The gender of the victim. Either M or F, string

- age: The age of the victim, integer

- race: The race of the victim. Either Asian/Pacific Islander, Native American/Native Alaskan, Black, Hispanic, or White, string

- hispanic: A code indicating the Hispanic origin of the victim, integer

- place: Where the shooting occurred. Has several categories, which you're encouraged to explore on your own, string

- education: Educational status of the victim, integer <br>
    can be one of the following: 
    - 1: Less than High School 
    - 2: Graduated from High School or equivalent 
    - 3: Some College 
    - 4: At least graduated from College 
    - 5: Not available

- id: Gun-death id, integer

** Simply run the following lines to connect to MongoDB database and import unicodecsv library **

In [1]:
import unicodecsv
from pymongo import MongoClient
from pprint import pprint
db = MongoClient()["Solvay"]
import pprint
pp = pprint.PrettyPrinter(depth=6)

### 4. Import data in Python

** 4.1 Open guns.csv file in a text editor (ex. Sublime Text) and identify "delimiter" and "quotechar" in the document**

quotechar:  
delimiter:

** 4.2 Explain what are delimiter and quotechar used for in a csv document **

delimiter in this case is a comma (,) and it is used to separate fields while the quotechar is a one-character string use to quote fields containing special characters. the defaults is ' " '

** 4.3 Import data from guns.csv in your Python notebook using DictReader. **

In [2]:
reader = unicodecsv.DictReader(open('./guns.csv'), quotechar='"', delimiter=',')

** 4.4 Print the five first lines of the csv files **

In [3]:
count = 1
for line in reader:
    if count <= 5:
        print(line)
    else:
        break
    count += 1

{u'education': u'4', u'police': u'0', u'age': u'34', u'sex': u'M', u'hispanic': u'100', u'race': u'Asian/Pacific Islander', u'intent': u'Suicide', u'year': u'2012', u'place': u'Home', u'month': u'1', u'id': u'1'}
{u'education': u'3', u'police': u'0', u'age': u'21', u'sex': u'F', u'hispanic': u'100', u'race': u'White', u'intent': u'Suicide', u'year': u'2012', u'place': u'Street', u'month': u'1', u'id': u'2'}
{u'education': u'4', u'police': u'0', u'age': u'60', u'sex': u'M', u'hispanic': u'100', u'race': u'White', u'intent': u'Suicide', u'year': u'2012', u'place': u'Other specified', u'month': u'1', u'id': u'3'}
{u'education': u'4', u'police': u'0', u'age': u'64', u'sex': u'M', u'hispanic': u'100', u'race': u'White', u'intent': u'Suicide', u'year': u'2012', u'place': u'Home', u'month': u'2', u'id': u'4'}
{u'education': u'2', u'police': u'0', u'age': u'31', u'sex': u'M', u'hispanic': u'100', u'race': u'White', u'intent': u'Suicide', u'year': u'2012', u'place': u'Other specified', u'month'

** 4.5 Which data structure is used by DictReader to import data from guns.csv ? **

In [10]:
type(line)

dict

Dictionary

** 4.6 DictReader uses the same data type to store all values. What is that data type ? **

string

### 5. Storing data in MongoDB

** 5.1 Which data structure will you use to store documents in MongoDB ? **

dictionary

** 5.2 Cite 2 advantages and 2 drawbacks of using document-oriented databases compared to relational databases ? **

Document-oriented databaseds allows for scalability over many servers and it has the dynamically add new attributes to data records.  These databases are disadvantages because they do not support ACID transactions. There is high chance of data duplicates 

Relational databases support ACID transactions and.....These databases are disadvantages as it has limitationsto store big data due to scalability issues on many servers. Takes time to design such databases and not flexible to schema changes. 

** 5.3 Identify an attribute in guns.csv that uniquely identifies each gun-death **

the field "id"

** 5.4 Store each gun-death information in a dedicated document in a MongoDB collection named "guns_death_us" **

For each gun-death document, store the following attributes in the specified data type:
- id: int
- year: int
- month: int
- intent: str
- police: int
- sex: str
- age: int
- race: str
- place: str
- education: int

In [5]:
def cast_int(arg):
    return int(arg)

In [4]:
reader = unicodecsv.DictReader(open('./guns.csv'), quotechar='"', delimiter=',')

In [8]:
for line in reader:
    line['id'] = cast_int(line['id'])
    line['year'] = cast_int(line['year'])
    line['month'] = cast_int(line['month'])
    line['police'] = cast_int(line['police'])  
    line['age'] = cast_int(line['age'])
    line['education']  = cast_int(line['education'])
    db['guns_death_us'].save(line)

** 5.5 Count how many documents are stored in newly created "guns_death_us" collection **

In [11]:
db['guns_death_us'].find().count()  

100719

### 6. Selecting documents from MongoDB

** 6.1 Select the first document in the "guns_death_us" collection **

In [12]:
db['guns_death_us'].find_one()

{u'_id': ObjectId('60466d2a01651c09e6b84eb1'),
 u'age': 50,
 u'education': 3,
 u'hispanic': u'100',
 u'id': 9,
 u'intent': u'Accidental',
 u'month': 2,
 u'place': u'Other specified',
 u'police': 0,
 u'race': u'White',
 u'sex': u'M',
 u'year': 2012}

** 6.2 Perform distinct operations on the following fields in order to identify their content: <br> **
- intent
- place
- race

"distinct" operation in MongoDB enables to identify all possible values in a field: **db["guns_death_us"].distinct('field') ** <br>

In [13]:
db["guns_death_us"].distinct('intent') 

[u'Accidental', u'Suicide', u'Homicide', u'Undetermined', u'NA']

In [14]:
db["guns_death_us"].distinct('place') 

[u'Other specified',
 u'Home',
 u'Street',
 u'NA',
 u'Other unspecified',
 u'Trade/service area',
 u'Farm',
 u'Industrial/construction',
 u'School/instiution',
 u'Sports',
 u'Residential institution']

In [15]:
db["guns_death_us"].distinct('race') 

[u'White',
 u'Native American/Native Alaskan',
 u'Asian/Pacific Islander',
 u'Black',
 u'Hispanic']

** 6.3 Select a gun-death that was accidental and happend in the street ** <br>
Only select age, intent, place, race and sex values

In [27]:
doc = db["guns_death_us"].find({"intent":"Accidental", "place":"Street"}, \
                               {"age":1,"intent":1,"race":1,"place":1, "race":1,"sex":1, "_id":0}).limit(3)

for elem in doc:
    pp.pprint(elem)

{u'age': 21,
 u'intent': u'Accidental',
 u'place': u'Street',
 u'race': u'Black',
 u'sex': u'M'}
{u'age': 26,
 u'intent': u'Accidental',
 u'place': u'Street',
 u'race': u'Hispanic',
 u'sex': u'M'}
{u'age': 64,
 u'intent': u'Accidental',
 u'place': u'Street',
 u'race': u'White',
 u'sex': u'M'}


** 6.4 Select gun-deaths that happened in a farm where homicide was commited on a woman ** <br>
Only select race and age values and print it for each document

In [26]:
docs = db['guns_death_us'].find({'place':'Farm','intent':'Homicide',"sex":"F"}).limit(3)

for elem in docs:
    pp.pprint(elem)

{u'_id': ObjectId('60466d2a01651c09e6b84f91'),
 u'age': 28,
 u'education': 2,
 u'hispanic': u'100',
 u'id': 234,
 u'intent': u'Homicide',
 u'month': 3,
 u'place': u'Farm',
 u'police': 0,
 u'race': u'White',
 u'sex': u'F',
 u'year': 2012}
{u'_id': ObjectId('60466d3201651c09e6b8c29d'),
 u'age': 27,
 u'education': 3,
 u'hispanic': u'210',
 u'id': 29704,
 u'intent': u'Homicide',
 u'month': 9,
 u'place': u'Farm',
 u'police': 0,
 u'race': u'Hispanic',
 u'sex': u'F',
 u'year': 2012}
{u'_id': ObjectId('60466d3401651c09e6b8d8d9'),
 u'age': 24,
 u'education': 1,
 u'hispanic': u'260',
 u'id': 35409,
 u'intent': u'Homicide',
 u'month': 5,
 u'place': u'Farm',
 u'police': 0,
 u'race': u'Hispanic',
 u'sex': u'F',
 u'year': 2013}


** 6.5  When selecting more than one document MongoDB returns a cursor and not a dictionary** <br>
- Describe what information does the cursor contain <br>
- Describe the purpose of this approach <br>

### 7. Exploring data

** 7.1  How many homicides were commited on Americans below 18 years old in 2014? ** <br>

In [28]:
db['guns_death_us'].find({'age':{'$lt':18},'year':2014}).count()

1322

** 7.2 Which month accounted for the highest number of gun-deaths in 2012 ? ** <br>
This question must be answered using an aggregation pipeline <br>
Sort results in a descending order <br>

In [29]:
db['guns_death_us'].aggregate([{"$group":{"_id":"$month","deaths":{"$sum":1}}},{"$sort":{"deaths":-1}}])

{u'ok': 1.0,
 u'result': [{u'_id': 7, u'deaths': 8983},
  {u'_id': 8, u'deaths': 8776},
  {u'_id': 6, u'deaths': 8672},
  {u'_id': 5, u'deaths': 8658},
  {u'_id': 9, u'deaths': 8503},
  {u'_id': 4, u'deaths': 8450},
  {u'_id': 12, u'deaths': 8409},
  {u'_id': 10, u'deaths': 8401},
  {u'_id': 3, u'deaths': 8283},
  {u'_id': 1, u'deaths': 8266},
  {u'_id': 11, u'deaths': 8235},
  {u'_id': 2, u'deaths': 7083}]}

** 7.3  Which intent caused the most gun-deaths in the US between 2012 and 2014 ? ** <br>
This question must be answered using an aggregation pipeline

In [39]:
db['guns_death_us'].aggregate([{"$match":{"year":{"$lte":2014,"$gte":2012}}}, \
                               {"$group":{"_id":"$intent","deaths":{"$sum":1}}},{"$sort":{"deaths":-1}}])

{u'ok': 1.0,
 u'result': [{u'_id': u'Suicide', u'deaths': 63155},
  {u'_id': u'Homicide', u'deaths': 35133},
  {u'_id': u'Accidental', u'deaths': 1625},
  {u'_id': u'Undetermined', u'deaths': 805},
  {u'_id': u'NA', u'deaths': 1}]}

** 7.4 Investigate sex distribution for the intent that caused the most gun-deaths in the US between 2012 and 2014 ? ** <br>
This question must be answered using an aggregation pipeline

In [59]:
db['guns_death_us'].aggregate([{"$match":{"year":{"$lte":2014,"$gte":2012}}}, \
                               {"$group":{"_id":{"intent": "$intent","sex":"$sex"}, "deaths":{"$sum":1}}}, \
                               {"$sort":{"sex":1,"deaths":-1, "intent":1}}])

{u'ok': 1.0,
 u'result': [{u'_id': {u'intent': u'Suicide', u'sex': u'M'}, u'deaths': 54469},
  {u'_id': {u'intent': u'Homicide', u'sex': u'M'}, u'deaths': 29777},
  {u'_id': {u'intent': u'Suicide', u'sex': u'F'}, u'deaths': 8686},
  {u'_id': {u'intent': u'Homicide', u'sex': u'F'}, u'deaths': 5356},
  {u'_id': {u'intent': u'Accidental', u'sex': u'M'}, u'deaths': 1410},
  {u'_id': {u'intent': u'Undetermined', u'sex': u'M'}, u'deaths': 636},
  {u'_id': {u'intent': u'Accidental', u'sex': u'F'}, u'deaths': 215},
  {u'_id': {u'intent': u'Undetermined', u'sex': u'F'}, u'deaths': 169},
  {u'_id': {u'intent': u'NA', u'sex': u'M'}, u'deaths': 1}]}

** 7.5 Draw a first conclusion on the type of gun-death that accounts for the most gun-deaths ? ** <br>

Suicide accounts for the most deaths (63,155) and the men (54,469) are the most affected in the US

** 7.6 Investigate race distribution for each of the top three intents ? ** <br>
This question must be answered using an aggregation pipeline <br>
Sort results in a descending order <br>

In [74]:
db['guns_death_us'].aggregate([{"$match":{"intent":"Suicide" }}, \
                               {"$group":{"_id":{"intent": "$intent","race":"$race"}, "deaths":{"$sum":1}}}, \
                               {"$sort":{"deaths":-1}}])

{u'ok': 1.0,
 u'result': [{u'_id': {u'intent': u'Suicide', u'race': u'White'},
   u'deaths': 55359},
  {u'_id': {u'intent': u'Suicide', u'race': u'Black'}, u'deaths': 3331},
  {u'_id': {u'intent': u'Suicide', u'race': u'Hispanic'}, u'deaths': 3169},
  {u'_id': {u'intent': u'Suicide', u'race': u'Asian/Pacific Islander'},
   u'deaths': 744},
  {u'_id': {u'intent': u'Suicide', u'race': u'Native American/Native Alaskan'},
   u'deaths': 552}]}

In [73]:
db['guns_death_us'].aggregate([{"$match":{"intent":"Homicide" }}, \
                               {"$group":{"_id":{"intent": "$intent","race":"$race"}, "deaths":{"$sum":1}}}, \
                               {"$sort":{ "deaths":-1}}])

{u'ok': 1.0,
 u'result': [{u'_id': {u'intent': u'Homicide', u'race': u'Black'},
   u'deaths': 19498},
  {u'_id': {u'intent': u'Homicide', u'race': u'White'}, u'deaths': 9125},
  {u'_id': {u'intent': u'Homicide', u'race': u'Hispanic'}, u'deaths': 5628},
  {u'_id': {u'intent': u'Homicide', u'race': u'Asian/Pacific Islander'},
   u'deaths': 557},
  {u'_id': {u'intent': u'Homicide',
    u'race': u'Native American/Native Alaskan'},
   u'deaths': 325}]}

In [72]:
db['guns_death_us'].aggregate([{"$match":{"intent":"Accidental" }}, \
                               {"$group":{"_id":{"intent": "$intent","race":"$race"}, "deaths":{"$sum":1}}}, \
                               {"$sort":{ "deaths":-1}}])

{u'ok': 1.0,
 u'result': [{u'_id': {u'intent': u'Accidental', u'race': u'White'},
   u'deaths': 1126},
  {u'_id': {u'intent': u'Accidental', u'race': u'Black'}, u'deaths': 321},
  {u'_id': {u'intent': u'Accidental', u'race': u'Hispanic'}, u'deaths': 145},
  {u'_id': {u'intent': u'Accidental',
    u'race': u'Native American/Native Alaskan'},
   u'deaths': 21},
  {u'_id': {u'intent': u'Accidental', u'race': u'Asian/Pacific Islander'},
   u'deaths': 12}]}

** 7.7 Describe impact of race on each of the top three intents ? ** <br>

Most deaths are registered by whites as they have the most deaths caused by suicide (55,359) and Accidental (1, 126). Blacks are mostly killed via Homicide (19, 498) followed by Whites (9,125). Hispanic and Native American/Native Alaskan are less affected as compared to the white and blacks. 

** 7.8 Investigate age distribution of victims on which were commited homicides ? ** <br>
This question must be answered using an aggregation pipeline <br>
Sort results in ascending order on age value <br>

In [77]:
db['guns_death_us'].aggregate([{"$match":{"intent":"Homicide" }}, \
                               {"$group":{"_id":"$age"}}, \
                               {"$sort":{ "age":-1}}])

{u'ok': 1.0,
 u'result': [{u'_id': 97},
  {u'_id': 91},
  {u'_id': 86},
  {u'_id': 94},
  {u'_id': 10},
  {u'_id': 77},
  {u'_id': 82},
  {u'_id': 89},
  {u'_id': 76},
  {u'_id': 107},
  {u'_id': 8},
  {u'_id': 11},
  {u'_id': 66},
  {u'_id': 21},
  {u'_id': 85},
  {u'_id': 58},
  {u'_id': 88},
  {u'_id': 4},
  {u'_id': 87},
  {u'_id': 83},
  {u'_id': 70},
  {u'_id': 6},
  {u'_id': 14},
  {u'_id': 35},
  {u'_id': 96},
  {u'_id': 67},
  {u'_id': 60},
  {u'_id': 64},
  {u'_id': 24},
  {u'_id': 56},
  {u'_id': 80},
  {u'_id': 37},
  {u'_id': 28},
  {u'_id': 15},
  {u'_id': 57},
  {u'_id': 81},
  {u'_id': 65},
  {u'_id': 74},
  {u'_id': 69},
  {u'_id': 79},
  {u'_id': 54},
  {u'_id': 1},
  {u'_id': 17},
  {u'_id': 95},
  {u'_id': 39},
  {u'_id': 51},
  {u'_id': 26},
  {u'_id': 12},
  {u'_id': 29},
  {u'_id': 16},
  {u'_id': 20},
  {u'_id': 50},
  {u'_id': 84},
  {u'_id': 45},
  {u'_id': 9},
  {u'_id': 42},
  {u'_id': 46},
  {u'_id': 13},
  {u'_id': 49},
  {u'_id': 90},
  {u'_id': 44},
  {u

** 7.9 Investigate average age of each race on which were commited at least 1000 homicides between 2012 and 2014 ? ** <br>
This question must be answered using an aggregation pipeline <br>
Sort results in a descending order <br>

In [21]:
db['guns_death_us'].aggregate([{"$match":{"year":{"$lte":2014,"$gte":2012}}},
                               {"$group":{"_id":{"race": "$race"}, "deaths":{"$sum":1},"avgAge":{"$avg":"$age"}}},
                               {"$match":{"deaths":{"$gte":1000}}},
                               {"$sort":{"avgAge":-1}}])

{u'ok': 1.0,
 u'result': [{u'_id': {u'race': u'White'},
   u'avgAge': 50.01776569227283,
   u'deaths': 66195},
  {u'_id': {u'race': u'Asian/Pacific Islander'},
   u'avgAge': 38.63492063492063,
   u'deaths': 1323},
  {u'_id': {u'race': u'Hispanic'},
   u'avgAge': 33.224206789438654,
   u'deaths': 9014},
  {u'_id': {u'race': u'Black'},
   u'avgAge': 31.119011815252417,
   u'deaths': 23275}]}

** 7.10 Describe impact of race on age of the victims of homicides ? ** <br>

Black race are victims of homicides at a younger age (31.1) on average as compared to other races. 

** 7.11 For each race, compute minimum age of the victim when police was involved in the gun-death ? ** <br>
This question must be answered using an aggregation pipeline <br>

In [25]:
db['guns_death_us'].aggregate([{"$match":{"police":1}},
                              {"$group":{"_id":{"race":"$race"},"minAge":{"$min":"$age"}}},
                              {"$sort":{"minAge":1}}])

{u'ok': 1.0,
 u'result': [{u'_id': {u'race': u'Black'}, u'minAge': 12},
  {u'_id': {u'race': u'Hispanic'}, u'minAge': 13},
  {u'_id': {u'race': u'White'}, u'minAge': 15},
  {u'_id': {u'race': u'Native American/Native Alaskan'}, u'minAge': 17},
  {u'_id': {u'race': u'Asian/Pacific Islander'}, u'minAge': 19}]}

** 7.12 Compare probability to be killed by accident between men and women victims ? ** <br>

In [28]:
db['guns_death_us'].aggregate([{"$match":{"intent":"Accidental"}},
                              {"$group":{"_id":{"sex":"$sex"},"deaths":{"$sum":1}}}])

{u'ok': 1.0,
 u'result': [{u'_id': {u'sex': u'F'}, u'deaths': 215},
  {u'_id': {u'sex': u'M'}, u'deaths': 1410}]}

In [38]:
prob_male = float(1410)/float((1410+215))
prob_female = float(215)/float((1410+215))
print prob_male 
print prob_female

0.867692307692
0.132307692308


** 7.13 Group gun-death by place-sex pair values and count the number of occurences for each pair value ** <br>
This question must be answered using an aggregation pipeline <br>
Sort results in a descending order <br>
Example of place-sex pair value: {'place': 'Home', 'sex': 'M'}

In [2]:
db['guns_death_us'].aggregate([{"$group":{"_id":{"place":"$place","sex":"$sex"},"Freq":{"$sum":1}}},
                              {"$sort":{"Freq":-1}}])

{u'ok': 1.0,
 u'result': [{u'Freq': 50180, u'_id': {u'place': u'Home', u'sex': u'M'}},
  {u'Freq': 12295, u'_id': {u'place': u'Other specified', u'sex': u'M'}},
  {u'Freq': 10310, u'_id': {u'place': u'Street', u'sex': u'M'}},
  {u'Freq': 10258, u'_id': {u'place': u'Home', u'sex': u'F'}},
  {u'Freq': 7657, u'_id': {u'place': u'Other unspecified', u'sex': u'M'}},
  {u'Freq': 3043, u'_id': {u'place': u'Trade/service area', u'sex': u'M'}},
  {u'Freq': 1442, u'_id': {u'place': u'Other specified', u'sex': u'F'}},
  {u'Freq': 1321, u'_id': {u'place': u'NA', u'sex': u'M'}},
  {u'Freq': 1200, u'_id': {u'place': u'Other unspecified', u'sex': u'F'}},
  {u'Freq': 837, u'_id': {u'place': u'Street', u'sex': u'F'}},
  {u'Freq': 561, u'_id': {u'place': u'School/instiution', u'sex': u'M'}},
  {u'Freq': 419, u'_id': {u'place': u'Farm', u'sex': u'M'}},
  {u'Freq': 396, u'_id': {u'place': u'Trade/service area', u'sex': u'F'}},
  {u'Freq': 228, u'_id': {u'place': u'Industrial/construction', u'sex': u'M'}},

** 7.14 Ask yourself a question about gun-deaths and find an answer by querying CDC dataset** <br>
Explain your question and how you answered it. 

Number of homicide per year?

In [3]:
db['guns_death_us'].aggregate([{"$group":{"_id":{"year":"$year"},"Total":{"$sum":1}}}])

{u'ok': 1.0,
 u'result': [{u'Total': 33584, u'_id': {u'year': 2014}},
  {u'Total': 33607, u'_id': {u'year': 2013}},
  {u'Total': 33528, u'_id': {u'year': 2012}}]}

** 7.15 Write a 10 lines conclusion on interesting insights that can be drawn from your exploration of the CDC dataset **

- Young black males are killed on average more than any other race
- Most of the homicides in the US occur at homes
- Women in general is less likely to killed in the US
- Generally White men are killed more in the US
- The amount of deaths from homicide for 2012 to 2014 was roughly thesame with 2013 been the highest